revamp auto-fill algorithm

lexelby opened the pull request:
This branch contains a rewrite of the auto-fill algorithm.  It's significantly simpler and faster, and I got rid of some of the "magic" heuristics.  It's a huge step up!

From the user perspective, not much will change.  But this new algorithm lays the groundwork for some huge new features (coming soon):

* Underpathing for travel stitches in fills (#257) 
  * the new algorithm seems to always do travel stitches before the rows they pass by, so we can offset them inside the shape
* Allow start or end for a fill to occur on the border of one of the holes, rather than the outer outline as is currently enforced.
* Curved fill and fill made of curlicues or other kinds of row shapes.

The code is also _much_ easier to understand, which is great from the perspective of attracting programmers to the project.

kaalleen commented:
Wow, so great!!! Thank you for your work on this!

lexelby commented:
I did it!!! This version does underpathing inside the fill region!

The path may look a little strange and jaggedy.  It's designed to avoid going in the same direction as the fill stitch so that it won't disrupt the look of the fill stitch that comes later.  I allow it to travel at a ±45 degree angle to the top stitching.  It uses a shortest-path algorithm to get from point A to point B, avoiding travel along the edges when possible.

Interestingly, this new algorithm is _tons_ faster than the old one.  The old one took great pains to avoid cutting corners when traveling around the outside of the shape.  That code was complicated and inefficient.  This one is way simpler and runs much faster.

You may notice that occasionally the algorithm underpaths even between two rows of stitching.  That's its way of avoiding the same problem, cutting a corner too close so that it would show up as an errant stitch outside the fill region.

To do:

* clean up the code a bit
* add an option to enable/disable underpathing separately for fill underlay and top stitching (default: enable)

@wwderw @kaalleen @X3msnake @jameskolme Really interested to hear what you folks think about this.

kaalleen commented:
First of all: this is really a great step forward!!

I created an odd shape and in the simulator it looked just great. I didn't notice what I could see in the print preview: sometimes the underpathing is stitched out on top of the fill and in one edge the underpathing is cutting corners.

![underpathing_on_top](https://user-images.githubusercontent.com/36401965/54418355-f0fc0400-4704-11e9-9c2c-8a5ab67749f8.jpg)

Here is the svg file: 
[underpath_on_top.zip](https://github.com/inkstitch/inkstitch/files/2970062/underpath_on_top.zip)

lexelby commented:
Aw darn, maybe I declared victory too early. ;)  I'll take a look at this later today and see if I can tweak the algorithm to prevent that.

jameskolme commented:
Yes! No more don´t go using that edge, because I can see it (not anyone else)  :) 

Similar errors with print preview rendering and noted maybe other issue, I can´t render individual parts. Every time complete design is rendered, not only individual design part.

lexelby commented:
Okay!  This version solves the problem with underpathing going on top of rows that have already been stitched.  Unfortunately, doing that is a bit computationally expensive, and it makes the algorithm take much longer to run.  Still shorter than the old one though, and I have some ideas on how to make it faster.

I haven't yet solved the issue with underpathing cutting the corner.  I'll tackle that next.

lexelby commented:
I found it!  It turns out that there's a special feature in Shapely that speeds this up a ton.  The slow part went from ~8 seconds to less than half a second!

Now we have a much simpler auto-fill algorithm that includes underpathing _and_ runs way faster than the old one.  Incredible!

Now for that corner-cutting issue.  I think I just figured out what's going on there.

lexelby commented:
Got it!  It was a tricky little bug.

I think that pretty much does it.  Let me know what you think.

> I created an odd shape 

Just to be clear here, I don't think that shape is odd at all. :)  We have to be able to support whatever kinds of shapes the user throws at us, so please do test it with all sorts of shapes.

lexelby commented:
The latest commit adds the ability to set the starting and/or ending point on the outline of one of the holes.  Previously this wasn't possible, and if you tried to do it, Ink/Stitch would just choose the closest point on the outer boundary.

There's one problem with this feature: Inkscape will still try to have the connector line point at the center of the shape.  If the hole is in the middle of the shape (like in a donut shape), then the connector will point to empty space.  You can move it manually, but any time you run any extension (e.g. Params), Inkscape will repoint it to the center of the shape.

It's possible to override this by deleting the `inkscape:connector-type` XML attribute on the connector line.  That disables Inkscape's auto-routing.  For the starting position, you can instead just put a tiny bit of running stitch at the spot you want to start at, just before the fill shape in the stacking order.  Ink/Stitch automatically defaults to a point closest to the last stitch if you don't specify a starting point for a fill.

In any case, this has bothered me in the past, so I'm glad to finally squash this one!

jameskolme commented:
There are now two check boxes on fill tab, I found it confusing :)

Then I went over the top for testing, underlay is cutting corners and fill seems to be little bit short on some places or underlay is out of line. End result is still amazing, because I tried some behind corner shape and it just goes everywhere.  Hmm maybe I should done some holes too. 

![underlaycornering](https://user-images.githubusercontent.com/44113605/54696821-31a8b280-4b35-11e9-851f-c61c47001d8f.png)

![big_fill](https://user-images.githubusercontent.com/44113605/54697023-8ba97800-4b35-11e9-8baa-bae73fb90508.png)

This error happens, if top underpath check box is selected first time
`Exception AttributeError: "'NoneType' object has no attribute 'GEOSSTRtree_destroy'" in <bound method STRtree.__del__ of <shapely.strtree.STRtree instance at 0x0D2BFD50>> ignored`

File:
[big_fill_test.zip](https://github.com/inkstitch/inkstitch/files/2988923/big_fill_test.zip)

lexelby commented:
Awesome shape:)

Oops! I forgot to mark the second checkbox as pertaining to underlay. I'll fix that this evening.

More corner cutting, hmm.  I see I may need to get more aggressive about preventing that.  I'll see what I can do.

lexelby commented:
As it turns out, the corner-cutting you're seeing in this case actually has nothing to do with the new auto-fill algorithm (:relieved:).  Even running stitch along a path like that would cut those corners too.  The fact is, the running stitch algorithm needs some work too, and it's on my list. :)

lexelby commented:
Regarding that exception, I can't seem to reproduce it.  I think I can write some defensive code to prevent that from happening, though.

lexelby commented:
Now I'm just getting fancy.  I tweaked the travel stitch pathfinder so that it prefers to travel in the center of the shape rather than close to the edge. :D

lexelby commented:
@wwderw Wanted to make sure you saw this.  I think it's a feature you've expressed interest in in the past.

wwderw commented:
> @wwderw Wanted to make sure you saw this. I think it's a feature you've expressed interest in in the past.

Oh yea.  I've been playing with it.  Just finished 3 sports themed stock designs using the dev version of this branch.

This improves the finished aesthetic when a fill doesn't have a satin outline around it's edge.  This allows for more options in the design.

I haven't had an issue with my final renders, but I didn't really stress it with the nice fill test that @jameskolme did, I used what I normally would stress it with a typical fill that I would deal with.

lexelby commented:
Just the fact that you're able to use Ink/Stitch in your professional work is really great :D

wwderw commented:
> Just the fact that you're able to use Ink/Stitch in your professional work is really great :D

Thanks for having a workable embroidery digitizing solution.

Having something that I could use for digitizing and run on a Linux system is no small task.

lexelby commented:
@catherineh here's the PR I was talking about :)

lexelby commented:
Another little enhancement: I added the ability to travel at 90 degrees to the fill stitch direction, so now paths are less jagged.

I also added a debug logging system that lets me see the networkx graphs I'm building as SVGs.  Should make debugging this kind of thing _much_ easier.

kaalleen commented:
Really nice work!
It seems to me, as if the checkbox for the underlay underpath has no effect. Everything is controlled by the top layer underpath checkbox (in the auto-fill tab) and cannot be handled independently.

lexelby commented:
Sounds like a copy-paste bug.  I'll work on it this evening hopefully.

kaalleen commented:
Great :)

There is something else, that was introduced by this branch. It can be easily removed (at least for this file) by simply moving the shape a tiny little bit. Any little changes would let this shape render without any problems, just not the way it is. I tried to reproduce the error and I am not able to. But luckily I saved the file, before I destroyed the error.

File: [fill-error](https://github.com/inkstitch/inkstitch/files/3025736/fill-error.svg.zip)

The error message:

> path815: error: Error during autofill!  This means that there is a problem with Ink/Stitch.
> 
> If you'd like to help us make Ink/Stitch better, please paste this whole message into a new issue at: https://github.com/inkstitch/inkstitch/issues/new
> 
>
> Traceback (most recent call last):
>   File "/home/momo/Projects/Manual-Inkstitch/inkstitch/lib/elements/auto_fill.py", line 210, in to_patches
>     self.underpath))
>   File "/home/momo/Projects/Manual-Inkstitch/inkstitch/lib/debug.py", line 125, in decorated
>     result = func(*args, **kwargs)
>   File "/home/momo/Projects/Manual-Inkstitch/inkstitch/lib/stitches/auto_fill.py", line 64, in auto_fill
>     travel_graph = build_travel_graph(fill_stitch_graph, shape, angle, underpath)
>   File "/home/momo/Projects/Manual-Inkstitch/inkstitch/lib/debug.py", line 125, in decorated
>     result = func(*args, **kwargs)
>   File "/home/momo/Projects/Manual-Inkstitch/inkstitch/lib/stitches/auto_fill.py", line 255, in build_travel_graph
>     p1, p2 = [InkstitchPoint(*coord) for coord in ls.coords]
> ValueError: too many values to unpack
>

lexelby commented:
Okay, I think that should take care of both problems.

kaalleen commented:
Yes it does, thanks for fixing! I just have to say it again... this PR is incredible and a huge step forward for Ink/Stitch ;)

lexelby commented:
I know, I'm so excited to release it!  This was a feature I was really dreading implementing, as I had no idea how to go about doing it.  It was really a stroke of luck that the new algorithm happened to work for underpathing!

I think we're pretty much ready to merge here.  Is anything else missing?

furicks commented:
I tested the latest release and it's definitely a big improvement in speed! The Underpathing works better but not quite perfect. Here's a quick test of the letter G you can see that the underpath is still on the perimeter in a few locations. 

![image](https://user-images.githubusercontent.com/5166589/55445446-726ae780-5589-11e9-8b4b-87a2e4ad65a4.png)

lexelby commented:
Thanks for testing!

There's a bit of a trade-off between perfection and speed going on here.  In my tests, it was doing pretty well most of the time, but I can see that in your case it didn't fare so well.

Would you be willing to share your SVG so that I can test with it?

furicks commented:
No problem! I really like the project, hopefully, I can contribute more! The existing commercial embroidery software is laughably outdated. 

[Groom Hat.zip](https://github.com/inkstitch/inkstitch/files/3036849/Groom.Hat.zip)

wwderw commented:
> The existing commercial embroidery software is laughably outdated.

In some areas I agree, but I think that's also the way it is with CAM software.  Because it also depends on hardware that changes glacially slow, that it seems to have carried over to the software side of things as well.

I'm just speculating as to why, but based on what I've seen with CAM software in general (from ones that communicate with embroidery machines, 3D Printers, CNC Routers etc) that seems to be the case.

lexelby commented:
@furicks We'd love to have your contributions!  There's lots to do for folks with all kinds of skills.  Translations, documentation, web design, and coding just to name a few.

Thanks very much for that example.  It helped me uncover and fix a bug that I hadn't been aware of.  The bug made it so that the auto-fill algorithm would incorrectly ignore certain sections of the interior of the shape when choosing a path for travel stitches, so it would fall back to going around the outside of the shape.  That one was easy to fix once I understood it.

There's another problem that's a little trickier.  Your shape is pretty small and that means that auto-fill has less space to construct its travel path.  Internally, auto-fill makes a grid like this:

![travel_mesh](https://user-images.githubusercontent.com/4446653/55595186-ae7c8480-5710-11e9-84a5-41be2188eb4d.png)

It uses the grid to construct a path from one section of the shape to the other.  It finds the shortest path from point A to point B, preferring paths that go inside the shape rather than around the outside border.

When your fill shape is big, it works really well because the grid is relatively small compared to the size of the shape.  When your shape is small like this G, auto-fill is more likely to have to briefly go around the outside border, like this:

![travel_border](https://user-images.githubusercontent.com/4446653/55595343-6873f080-5711-11e9-9a99-257609c7e57f.png)

If you want truly "perfect" underpathing, it's necessary to make the grid much finer.  That can mean it takes much longer for the auto-fill algorithm to do its work.

If you really care about even short runs like this, I think the best way to handle it would be to add a checkbox:

  ☒ try harder to avoid traveling around the border

The tooltip for that would explain that it's a trade-off: if you tell it to try harder it'll take longer.

What do you think?

CatherineH commented on lib/debug.py:
```diff
@@ -0,0 +1,197 @@
+
+    @check_enabled
+    def log_svg_element(self, element):
+        if self.current_layer is None:
```
I think it's more pythonic to check for none without the ``is None``` unless self.current_layer will sometimes be 0 or an empty string

lexelby replied:
Oh, this one's a little tricky.  `self.current_layer` is an lxml element tag.  Element instances act like lists and you can index them, slice, append, etc.  

If you try to treat one as a boolean directly, it complains at you with a deprecation warning because it's ambiguous whether you mean None or an XML tag with no child tags.  So we have to compare to None in this case.

CatherineH commented on lib/debug.py:
```diff
@@ -0,0 +1,197 @@
+        path = line_strings_to_path(line_strings)
+        path.set('style', formatStyle({"stroke": color or "#000000", "stroke-width": "0.3"}))
+
+        if name is not None:
```
I think
```
if not name
```

is more pythonic

lexelby replied:
I could do `if name`, but that would be false for an empty string.  I guess it arguably isn't all that useful to set an empty string as a label, but I don't see why it shouldn't be possible.

I generally seem to like to compare things with `is None` when I know the thing might be None just to make it clear what I'm doing.

All that is to say I'm on the fence on this one, but I think I'll leave it.

CatherineH commented on lib/debug.py:
```diff
@@ -0,0 +1,197 @@
+        }))
+
+    @contextmanager
+    def time_this(self, label="code block"):
```
does this context ever get applied when self.enabled is False?

lexelby replied:
Yup!  My idea with this class is that I can add debug statements to the code and leave them in.  When debugging is disabled, the debug statements should have minimal impact.  For this context manager, it should just be a noop if debug is disabled.

Previously I was adding and removing debug statements whenever I needed them.  Especially for things like timing code, it was getting tedious.

CatherineH commented on lib/svg/__init__.py:
```diff
@@ -1,4 +1,4 @@
+from .guides import get_guides
+from .path import apply_transforms, get_node_transform, get_correction_transform, line_strings_to_csp, point_lists_to_csp, line_strings_to_path
```
Do you lint your inits or imports? I would make this newline two lines

lexelby replied:
I do!  See the `style` target in the Makefile.  I chose to set the maximum line length to 150, since 80 columns seems too restrictive given today's monitor resolutions.

FWIW, I use Eclipse with PyDev and have it set to run `autopep8` whenever I save.  It sorts my imports and splits long lines automatically.

CatherineH commented on lib/stitches/auto_fill.py:
```diff
@@ -54,18 +56,17 @@ def auto_fill(shape,
-
-    graph = build_graph(shape, segments, angle, row_spacing, max_stitch_length)
-    path = find_stitch_path(graph, segments, starting_point, ending_point)
+    fill_stitch_graph = build_fill_stitch_graph(shape, angle, row_spacing, end_row_spacing)
```
This block is kind of hard to read... do you see any value in turning this into a namespace or dataclass so that you don't have to pass so many input variables around?

lexelby replied:
I've kind of been on the fence about that.  On the one hand, it would make the function calls less cumbersome -- and I do seem to keep adding more and more parameters as time goes on.

On the other hand, I'm not sure what kind of object the class would represent.  An "autofiller"?

If you want to take a crack at it, I'd like to see what you come up with.  You're probably right that it would be tidier.  The same would apply to `lib/stitches/auto_satin.py`.

lexelby replied:
I think for now I'm going to leave it as it is, since this PR is already getting pretty big.  I'd be open to trying your idea in a separate PR.

CatherineH commented on lib/stitches/auto_fill.py:
```diff
@@ -128,248 +131,241 @@ def build_graph(shape, segments, angle, row_spacing, max_stitch_length):
 
-def find_loop(graph, starting_nodes):
-    """find a loop in the graph that is connected to the existing path
+    if underpath:
```
I would return early in this function so that the rest of the code isn't indented, i.e., 

```
if not underpath:
     return
```

lexelby replied:
Good idea.  This function got longer than it really should be.  I think I'll try to split it out a bit more and return early as you suggest.

CatherineH commented on lib/stitches/auto_fill.py:
```diff
@@ -128,248 +131,241 @@ def build_graph(shape, segments, angle, row_spacing, max_stitch_length):
-    else:
-        outline_nodes = get_outline_nodes(graph)
+def check_graph(graph, shape, max_stitch_length):
+    if networkx.is_empty(graph) or not networkx.is_eulerian(graph):
```
I would write:

```
error_msg = "This shape is so small that it cannot be filled with rows of stitches.  "
                     "It would probably look best as a satin column or running stitch."
                      if shape.area < max_stitch_length ** 2 else
                      "Cannot parse shape. This most often happens because your shape is "
                      "made up of multiple sections that aren't connected."
raise InvalidPath(_(error_msg))
```

lexelby replied:
Done, thanks.

CatherineH commented on lib/stitches/auto_fill.py:
```diff
@@ -392,51 +388,81 @@ def find_stitch_path(graph, segments, starting_point=None, ending_point=None):
-    # That's how we force the starting and ending point.
-
-    nodes_visited.append(path[0][0])
+    if ending_point is None:
```
```
if not ending_point
```

CatherineH commented on lib/stitches/auto_fill.py:
```diff
@@ -392,51 +388,81 @@ def find_stitch_path(graph, segments, starting_point=None, ending_point=None):
+    while vertex_stack:
+        current_vertex, current_key = vertex_stack[-1]
+        if graph.degree(current_vertex) == 0:
+            if last_vertex is not None:
```
is not None probably not needed

CatherineH commented on lib/stitches/auto_fill.py:
```diff
@@ -392,51 +388,81 @@ def find_stitch_path(graph, segments, starting_point=None, ending_point=None):
+    # Note, it's quite possible that part of this PathEdge will be eliminated by
+    # collapse_sequential_outline_edges().
+
+    if starting_node is not ending_node:
```
Do you mean to compare identity or do you want an inequality here?

lexelby replied:
I really want identity.  This case tests for the situation where I set `ending point = starting_point` above (when `ending_point` was not passed and defaulted to None).  I'm being perhaps overly cautious to avoid comparing floating point numbers with all the complication that can entail.

lexelby replied:
Oh, and an identity test is faster, I suppose.

CatherineH commented on lib/stitches/auto_fill.py:
```diff
@@ -392,51 +388,81 @@ def find_stitch_path(graph, segments, starting_point=None, ending_point=None):
+    # relevant in the case that the user specifies an underlay with an inset
+    # value, because the starting point (and possibly ending point) can be
+    # inside the shape.
+    outline_nodes = [node for node, outline in travel_graph.nodes(data="outline") if outline is not None]
```
```
if outline]
```

lexelby replied:
`outline` is a zero-indexed number in this case.  For once my comparison to None is valid! :tada:

lexelby commented:
YAY! Thank you!  I've been wishing for someone to review my code.  Style and understandability are really important to me, so that we can bring in more devs like you.  See also: `make style`, enforced by the Travis build.

I'm a bit busy this weekend but I'll try to go through your comments soon.

lexelby commented:
Okay, I think we're ready to go here. :D  I'm going to merge and make a new release.

@furicks, I'm still interested to hear what you think regarding the above.  If we do decide to add a "try harder" checkbox, I'll do it in a later release.

furicks commented:
I like the try harder checkbox concept. I think it's fine if processing takes a while, but the addition of a progress bar of some kind would be nice if the processing times would increase. I know I was confused if anything was happening when I first installed the program. 

Logically it seems like the same grid plus a crude offset would solve the problem entirely. Follow a grid but don't get any more than 1mm (parameter?) from the edge of the shape. If the shape is too small then go with the shortest path.

lexelby commented:
Yeah, I totally agree about a progress bar.  I'm hoping that our current work on revamping the UI (#426) might make something like that possible.
